minio/minio
# Kafka notify: support batched commits for queue store

## Community Contribution License
All community contributions in this pull request are licensed to the project maintainers
under the terms of the [Apache 2 license](https://www.apache.org/licenses/LICENSE-2.0). 
By creating this pull request I represent that I have the right to license the 
contributions to the project maintainers under the Apache 2 license.

## Description

The items will be saved in per target batch and will be committed to the queue store when batch is full.
Also peridically commits the batched items to the queue stored based on configured commit_timeout; default is 30s;
    
bonus: compress queue store multiwrites

have added the support for `notify_kafka` for now, will add the support for the rest later

## Motivation and Context

To reduce the queue_store IO calls and compress multiple event payloads as these payloads are compressible candidates.

## How to test this PR?

- Setup Kafka - Refer https://kafka.apache.org/quickstart
- Enable `notify_kafka` with `batch_size`
```
mc admin config set myminio notify_kafka enable=on topic=quickstart-events brokers=localhost:9092 queue_dir=/tmp/events batch_size=<n>
```
- Add bucket notification using `mc event add` and make some operations on the bucket
- Bring down the target and you will see the events persist in batches. Like the following

``` sh
> tree -a /tmp/events/
/tmp/events/
└── minio-kafka-_
    ├── 23:72c3fa71-8d55-4da0-bb68-a3ed37e26fb1.event.snappy
    └── 37:3b1d50e9-bd5c-4eaa-9170-22f3a1e911aa.event.snappy
```

- Bring up the kafka target after sometime, and notice the events getting replayed in batches
- After bringing the target up, there shouldn't be any messages left in the queue store - `ls -l /tmp/events`

## Types of changes

## Checklist:
